tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,747 @@
+    /// therefore other references can exist to it while mutable references to
+    /// Entry exist.
+    ///
+    /// This is manipulated only under the inner mutex. TODO: Can we use loom
+    /// cells for this?
+    inner: StdUnsafeCell<TimerShared>,
+}
 
-    pub(crate) fn fire(&self, when: u64) {
-        let mut curr = self.state.load(SeqCst);
+unsafe impl<TS: TimeSource> Send for TimerEntry<TS> {}
+unsafe impl<TS: TimeSource> Sync for TimerEntry<TS> {}
 
-        loop {
-            if is_elapsed(curr) || curr > when {
-                return;
-            }
+/// An EntryHandle is the (non-enforced) "unique" pointer from the driver to the
+/// timer entry. Generally, at most one EntryHandle exists for a timer at a time
+/// (enforced by the timer state machine).
+///
+/// SAFETY: An EntryHandle is essentially a raw pointer, and the usual caveats
+/// of pointer safety apply. In particular, EntryHandle does not itself enforce
+/// that the timer does still exist; however, normally an EntryHandle is created
+/// immediately before registering the timer, and is consumed when firing the
+/// timer, to help minimize mistakes. Still, because EntryHandle cannot enforce
+/// memory safety, all operations are unsafe.
+#[derive(Debug)]
+pub(crate) struct TimerHandle {
+    inner: NonNull<TimerShared>,
+}
 
-            let next = ELAPSED | curr;
-            let actual = self.state.compare_and_swap(curr, next, SeqCst);
+pub(super) type EntryList = crate::util::linked_list::LinkedList<TimerShared, TimerShared>;
 
-            if curr == actual {
-                break;
-            }
+/// The shared state structure of a timer. This structure is shared between the
+/// frontend (`Entry`) and driver backend.
+///
+/// Note that this structure is located inside the `TimerEntry` structure.
+#[derive(Debug)]
+pub(crate) struct TimerShared {
+    /// Current state. This records whether the timer entry is currently under
+    /// the ownership of the driver, and if not, its current state (not
+    /// complete, fired, error, etc).
+    ///
+    /// This value is accessed atomically from multiple threads and so some care
+    /// is needed to use this properly. The general rules are:
+    ///
+    /// 1. If the state is not Registered or FiringInProgress, the driver does
+    ///    not own the timer and is not allowed to touch it at all. As such, the
+    ///    timer future (Entry) is free to manipulate the timer freely.
+    /// 2. Entering or leaving Registered or FiringInProgress requires the
+    ///    driver lock. Entering Registered further requires exclusive ownership
+    ///    of the timer future.
+    /// 3. The state of a timer can only be FiringInProgress while the driver
+    ///    lock is held by the driver itself.
+    ///
+    /// It follows that cancellation of a timer requires acquiring the driver
+    /// lock if the state is Registered or FiringInProgress.
+    state: state_cell::StateCell,
```

Comment:
nit: there's a `use` for this above
```suggestion
    state: StateCell,
```